File: days/day20.swift
#!/usr/bin/env swift
import Foundation


struct V2 : Hashable {
    var x: Int
    var y: Int

    static let up = V2(x: 0, y: -1)
    static let down = V2(x: 0, y: 1)
    static let left = V2(x: -1, y: 0)
    static let right = V2(x: 1, y: 0)
    static let around = [V2.right, V2.down, V2.left, V2.up]

    static func + (a: Self, b: Self) -> V2 {
        V2(x: a.x + b.x, y: a.y + b.y)
    }
}


struct Problem {
    var grid: Set<V2>
    var start: V2
    var goal: V2
}


func parseProblem(_ data: String) -> Problem {
    let grid: [V2:Character] = data
        .split(separator: "\n")
        .enumerated()
        .flatMap { p in Array(p.element).enumerated().compactMap { q in
            q.element == "#" ? nil :
                (q.element, V2(x: q.offset, y: p.offset)) } }
        .reduce(into: [:], { acc, p in acc[p.1] = p.0 })
    let start = grid.first(where: { $0.value == "S" })!.key
    let goal = grid.first(where: { $0.value == "E" })!.key
    return Problem(grid: Set(grid.keys), start: start, goal: goal)
}


func part1(_ data: String, n: Int = 100, t: Int = 2) -> Int {
    let prob = parseProblem(data)

    func findPaths(_ start: V2) -> [V2:Int] {
        var fringe: [(Int,V2)] = [(0, start)]
        var seen: [V2:Int] = [:]

        while let (distance, pos) = fringe.first {
            fringe.removeFirst()
            if seen.keys.contains(pos) {
                continue
            }
            seen[pos] = distance

            for d in V2.around {
                if prob.grid.contains(pos + d) {
                    fringe.append((distance + 1, pos + d))
                }
            }
        }

        return seen
    }

    let from_start = findPaths(prob.start)
    let to_end = findPaths(prob.goal)
    let base = from_start[prob.goal]!
    var ans = 0

    for (pos, dt) in from_start {
        for v in (1 ... t) {
            for x in (-v ... v) {
                for y in [v - abs(x), -v + abs(x)] {
                    let q = pos + V2(x: x, y: y)
                    if let w = to_end[q] {
                        let ok = base - (dt + v + w) >= n
                        ans += ok ? 1 : 0
                    }
                    if y == 0 { break }
                }
            }
        }
    }

    return ans
}


func part2(_ data: String, n: Int = 100, t: Int = 20) -> Int {
    part1(data, n: n, t: t)
}


let test = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""
assert(part1(test, n: 38) == 3)
assert(part2(test, n: 76) == 3)


let data = try! String(contentsOfFile: "day20.in", encoding: .utf8)
let ans1 = part1(data)
print("part1: \(ans1)")
let ans2 = part2(data)
print("part2: \(ans2)")

File: days/day20.py
#!/usr/bin/env python
from collections import deque


def part1(data, N=100, T=2):
    lines = data.strip().splitlines()
    w, h = len(lines[0]), len(lines)
    grid = {(x + 1j*y):c for y,s in enumerate(lines) for x,c in enumerate(s)
        if c != '#'}
    start = next(p for p, c in grid.items() if c == 'S')
    goal = next(p for p, c in grid.items() if c == 'E')
    neib = [1, 1j, -1, -1j]

    def find_paths(start):
        fringe = deque([(0, start)])
        seen = dict()
        while fringe:
            dt, pos = fringe.popleft()
            if pos in seen: continue
            seen[pos] = dt
            for d in neib:
                if (q := pos + d) in grid:
                    fringe.append((dt + 1, q))
        return seen

    from_start = find_paths(start)
    to_end = find_paths(goal)
    base = from_start[goal]
    ans = 0

    for pos, dt in from_start.items():
        for n in range(1, T+1):
            for x in range(-n, n+1):
                for y in [n - abs(x), -n + abs(x)]:
                    q = pos + (x + 1j * y)
                    if (w := to_end.get(q)) is not None:
                        ans += base - (dt + n + w) >= N
                    if not y: break
    return ans


def part2(data, N=100):
    return part1(data, N, T=20)


data = '''
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
'''
assert part1(data, N=38) == 3
assert part2(data, N=76) == 3


data = open('day20.in').read()
ans = part1(data)
print('part1:', ans)
ans = part2(data)
print('part2:', ans)
